include/clang/AST/Expr.h

```diff
@@ -955,9 +955,43 @@ class DeclRefExpr final
       private llvm::TrailingObjects<DeclRefExpr, NestedNameSpecifierLoc,
                                     NamedDecl *, ASTTemplateKWAndArgsInfo,
                                     TemplateArgumentLoc> {
+public :
+  /// \brief extra information for DeclRefExprInfo.
+  class GenericFunctionCallInfo {
+  public:
+    struct TypeNameInfo {
+      QualType typeName;
+      TypeSourceInfo *sourceInfo;
+    };
+
+  private :
+    /// \brief If DeclRefExpr references generic function, it must store
+    /// extra information about type names for generic function initialization
+    TypeNameInfo *TypeNameInfos;
+    unsigned int NumTypeNameInfo;
+
+    GenericFunctionCallInfo()
+      : TypeNameInfos(nullptr), NumTypeNameInfo(0) {}
+
+  public :
+    static GenericFunctionCallInfo *Create(ASTContext &C,
+      ArrayRef<TypeNameInfo> NewTypeVariableNames);
+
+    ArrayRef<TypeNameInfo> typeNameInfos() const {
+      return{ TypeNameInfos, NumTypeNameInfo };
+    }
+
+    MutableArrayRef<TypeNameInfo> typeNameInfos() {
+      return{ TypeNameInfos, NumTypeNameInfo };
+    }
+  };
+
+private :
   /// \brief The declaration that we are referencing.
   ValueDecl *D;
 
+  GenericFunctionCallInfo *GenFuncInfo;
```

Comment:
I suggest renaming this to something like TypeArguments.   Please add a comment describing this field.